File: tests/test_startup.py
"""Process startup: a bad environment must fail cleanly, not traceback."""

from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path

from conftest import REPO_ROOT

ENTRY = [sys.executable, "-m", "cogball.server"]


def run(env_extra: dict, timeout: int = 30):
    env = dict(os.environ)
    env["PYTHONPATH"] = os.pathsep.join(
        [str(REPO_ROOT / "server"), str(REPO_ROOT)])
    env.pop("COGAME_CONFIG_URI", None)
    env.pop("COGAME_LOAD_REPLAY_URI", None)
    env.update(env_extra)
    return subprocess.run(ENTRY, capture_output=True, text=True, env=env,
                          cwd=str(REPO_ROOT), timeout=timeout)


def test_a_missing_config_uri_exits_2_with_a_clean_message():
    proc = run({})
    assert proc.returncode == 2
    assert "COGAME_CONFIG_URI is required" in proc.stderr
    assert "Traceback" not in proc.stderr


def test_an_unreadable_config_exits_2_with_a_clean_message(tmp_path):
    proc = run({"COGAME_CONFIG_URI": f"file://{tmp_path}/nope.json"})
    assert proc.returncode == 2
    assert "invalid config" in proc.stderr
    assert "Traceback" not in proc.stderr


def test_config_that_is_not_json_exits_2_with_a_clean_message(tmp_path):
    path = Path(tmp_path) / "config.json"
    path.write_text("{not json")
    proc = run({"COGAME_CONFIG_URI": f"file://{path}"})
    assert proc.returncode == 2
    assert "not valid JSON" in proc.stderr
    assert "Traceback" not in proc.stderr


def test_a_config_with_the_wrong_seat_count_exits_2(tmp_path):
    path = Path(tmp_path) / "config.json"
    path.write_text(json.dumps({
        "players": [{"name": "a"}, {"name": "b"}, {"name": "c"}],
        "tokens": ["t0", "t1", "t2"]}))
    proc = run({"COGAME_CONFIG_URI": f"file://{path}"})
    assert proc.returncode == 2
    assert "seats exactly 2 players" in proc.stderr
    assert "Traceback" not in proc.stderr


def test_num_agents_disagreeing_with_the_seat_count_exits_2(tmp_path):
    path = Path(tmp_path) / "config.json"
    path.write_text(json.dumps({
        "players": [{"name": "a"}, {"name": "b"}],
        "tokens": ["t0", "t1"], "num_agents": 6}))
    proc = run({"COGAME_CONFIG_URI": f"file://{path}"})
    assert proc.returncode == 2
    assert "num_agents 6" in proc.stderr
    assert "Traceback" not in proc.stderr


def test_a_corrupt_replay_in_replay_mode_exits_2(tmp_path):
    path = Path(tmp_path) / "replay.json"
    path.write_bytes(b"{\"protocol\": \"nope\"}")
    proc = run({"COGAME_LOAD_REPLAY_URI": f"file://{path}"})
    assert proc.returncode == 2
    assert "invalid replay" in proc.stderr
    assert "Traceback" not in proc.stderr


def test_the_player_entry_point_fails_cleanly_with_no_url():
    env = dict(os.environ)
    env["PYTHONPATH"] = os.pathsep.join(
        [str(REPO_ROOT / "server"), str(REPO_ROOT)])
    env.pop("COWORLD_PLAYER_WS_URL", None)
    env.pop("COGAMES_ENGINE_WS_URL", None)
    proc = subprocess.run(
        [sys.executable, "-m", "players.cogball_player"],
        capture_output=True, text=True, env=env, cwd=str(REPO_ROOT),
        timeout=30)
    assert proc.returncode == 1
    assert "no websocket URL" in proc.stderr
    assert "Traceback" not in proc.stderr


def test_the_image_declares_both_role_shims():
    """The two /bin entrypoints the manifest and policies.json name.

    The docker smoke proves they are executable inside the real image; this
    keeps the Dockerfile and the manifest from drifting apart in a diff.
    """
    dockerfile = (REPO_ROOT / "Dockerfile").read_text()
    assert "/bin/cogball" in dockerfile
    assert "/bin/cogball-player" in dockerfile
    assert "chmod +x /bin/cogball /bin/cogball-player" in dockerfile
    assert 'CMD ["/bin/cogball"]' in dockerfile

File: tests/test_startup.nim
## Startup contract: the entrypoint dies with a clean message and no traceback
## on a bad config, the seed is randomised when unpinned and honoured when
## pinned, and both entrypoints exist in the image.

import std/[os, strutils]
import lib/helpers
import cogball/sim_config

proc badConfigIsRejectedCleanly() =
  ## `config.update` raises a CogballError with a readable message; the
  ## entrypoint turns that into `quit(msg, 1)`, never a traceback.
  var config = defaultGameConfig()
  var message = ""
  try:
    config.update("this is not json")
  except CogballError as error:
    message = error.msg
  doAssert message.contains("not valid JSON"),
    "an unparseable config did not say so: " & message
  message = ""
  try:
    var other = defaultGameConfig()
    other.update("[1,2,3]")
  except CogballError as error:
    message = error.msg
  doAssert message.contains("must be a JSON object"), message
  message = ""
  try:
    var other = defaultGameConfig()
    other.update("""{"num_agents": 6}""")
  except CogballError as error:
    message = error.msg
  doAssert message.contains("num_agents must be 2"), message
  message = ""
  try:
    var other = defaultGameConfig()
    other.update("""{"maxTicks": 0}""")
  except CogballError as error:
    message = error.msg
  doAssert message.contains("maxTicks"), message
  message = ""
  try:
    var other = defaultGameConfig()
    other.update("""{"turnBudgetMs": 100}""")
  except CogballError as error:
    message = error.msg
  doAssert message.contains("turnBudgetMs"), message
  # The entrypoint's own handling: a clean message, exit 1, no traceback.
  let source = readFile("src/cogball.nim")
  doAssert source.contains("quit(\"cogball: bad config: \" & error.msg, 1)"),
    "the entrypoint no longer turns a bad config into a clean exit"
  report "a bad config is rejected with a clean message and exit 1"

proc seedIsPinnedOrRandomised() =
  ## The compiled-in default doubles as the "nobody chose a seed" sentinel, and
  ## it is deliberately NOT the certification fixture's seed.
  doAssert DefaultSeed != 679961,
    "the default seed collides with the certification fixture's pin"
  let source = readFile("src/cogball.nim")
  doAssert source.contains("const LegacyFixedSeed = DefaultSeed")
  doAssert source.contains("config.seed = randomSeed()")
  doAssert source.contains("stripUnpinnedSeed"),
    "an unpinned sentinel seed could clobber the randomised one"
  # A pinned seed is honoured end to end.
  var config = defaultGameConfig()
  config.update("""{"seed": 424242}""")
  doAssert config.seed == 424242
  var sim = initSimServer(config)
  doAssert sim.config.seed == 424242
  # ...and it determines the kickoff.
  var a = seatedSim(testConfig(seed = 4))
  var b = seatedSim(testConfig(seed = 4))
  a.startGame()
  b.startGame()
  for i in 0 ..< RobotCount:
    doAssert a.robots[i].y == b.robots[i].y
  var c = seatedSim(testConfig(seed = 5))
  c.startGame()
  var different = false
  for i in 0 ..< RobotCount:
    if c.robots[i].y != a.robots[i].y:
      different = true
  doAssert different, "two different seeds produced the same kickoff"
  report "the seed is randomised when unpinned and honoured when pinned"

proc bothEntrypointsAreDeclared() =
  ## The docker smoke asserts they EXIST in the image; here we assert the repo
  ## still declares both and that they come out of one image.
  doAssert fileExists("src/cogball.nim")
  doAssert fileExists("src/cogball_player.nim")
  let dockerfile = readFile("Dockerfile")
  doAssert dockerfile.contains("--out:cogball src/cogball.nim")
  doAssert dockerfile.contains("--out:cogball-player src/cogball_player.nim")
  doAssert dockerfile.contains("COPY --from=build /workspace/cogball/cogball /bin/cogball")
  doAssert dockerfile.contains("/bin/cogball-player")
  doAssert dockerfile.contains("COPY --from=build /workspace/cogball/data ./data"),
    "the runtime stage does not ship data/ — the rig art would be missing"
  doAssert dockerfile.contains("CMD [\"/bin/cogball\"]")
  let player = readFile("src/cogball_player.nim")
  doAssert player.contains("COWORLD_PLAYER_WS_URL")
  doAssert player.contains("PLAYER_PROMPT")
  doAssert player.contains("PLAYER_SCRIPTED")
  doAssert player.contains("PLAYER_POLICY_LABEL")
  report "one image declares both entrypoints and ships the art"

proc theRuntimeContractIsRead() =
  let source = readFile("src/cogball/server.nim")
  for name in ["COGAME_EVENTS_URI", "COGAME_PLAYER_FAILURE_URI"]:
    doAssert source.contains(name), "the server no longer reads " & name
  let entry = readFile("src/cogball.nim")
  doAssert entry.contains("readRuntimeConfig()"),
    "the entrypoint no longer reads the COGAME_* runtime contract"
  report "the COGAME_* runtime contract is read at startup"

proc missingConfigUriIsSurvivable() =
  ## No COGAME_CONFIG_URI at all is the DEFAULT config, not a crash: the smoke
  ## and the certifier both rely on that.
  var config = defaultGameConfig()
  config.update("")
  doAssert config.numAgents == 2
  doAssert config.maxTicks == DefaultMaxTicks
  report "an absent config falls back to the defaults, cleanly"

when isMainModule:
  echo "test_startup"
  badConfigIsRejectedCleanly()
  seedIsPinnedOrRandomised()
  bothEntrypointsAreDeclared()
  theRuntimeContractIsRead()
  missingConfigUriIsSurvivable()
  echo "test_startup: all good"
